active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,206 @@
+* Server MAY ask the client to abort file transfer
+* The protocol MUST guarantee "At least once" delivery
+* Server MUST NOT support subscription on topics dedicated for file transfer
+
+## AWS IoT MQTT-based file delivery (reference design)
+
+As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):
+
+![](0021-assets/aws-mqtt-file-delivery.png)
+
+## Design
+
+### Overview
+
+* Files are split in segments, segments can be of arbitrary length
+* Client generates UUID for each file being transferred and use it as file Id in Topic Name
+* Client calculates sha256 checksum of the segment it's about to send and sends it as part of Topic Name
+* Client uses $file Topic Filter to transfer files
+* Clients cannot subscribe to $file topics
+* Segment length can be calculated on the server side by subtracting the length of the [Variable Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901025) from the [Remaining Length](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901105) field that is in the [Fixed Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901021)
+* Data is transferred in PUBLISH packets in the following order:
+  1. $file/{fileId}/init
+  2. $file/{fileId}/{offset}/{sha256sum}
+  3. $file/{fileId}/{offset}/{sha256sum}
+  4. ...
+  3. $file/{fileId}/[fin/{sha256sum} | abort]
+* Client can send up to N (configured) PUBLISH packets before blocking for PUBACK
+
+#### `$file/{fileId}/init` message
+
+Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file metadata.
+
+  * Qos=1
+  * DUP=0 for the initial segment transfer, 1 when retransmitting
+  * Payload Format Indicator=0x01
+  * `{fileId}` is corresponding file UUID
+  * Payload is a JSON document with the following properties:
+    * filepath
+      * a character string composed with characters from this set [a-zA-Z0-9_-./]
+      * length in bytes MUST be <= 4096 (PATH_MAX=4096 on most common Linux filesystems)
+    * filename
+      * a character string composed with characters from this set [a-zA-Z0-9_-./]
+      * length in bytes MUST be < 256 (NAME_MAX=255 in Linux)
+    * size [optional]
+      * total file size in bytes
+
+#### `$file/{fileId}/{offset}/{sha256sum}` message
+
+One such message for each file segment.
+
+  * Qos=1
+  * DUP=0 for the initial segment transfer, 1 when retransmitting
+  * Payload Format Indicator=0x00
+  * Packet Identifier=sequential file segment number
+  * Payload is file segment bytes
+  * `{offset}` is byte offset of the given segment
+  * `{sha256sum}` is sha256 checksum of file segment bytes
```

Comment:
* can we have offset and sha256 as part of the payload?
** upside: static and short topic names
** upside: clients can use topic alias if they support it
** downside: clients have to assemble payload, we need a serialization format